src/transformers/data/data_collator.py

```diff
@@ -134,3 +134,107 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
+    max_gram: int = 5  # maximum individual mask span length L
+
+    def __call__(self, examples: List[torch.Tensor]) -> Dict[str, torch.Tensor]:
+        batch = self._tensorize_batch(examples)
+        inputs, perm_mask, target_mapping, labels = self.mask_tokens(batch)
+        return {"input_ids": inputs, "perm_mask": perm_mask, "target_mapping": target_mapping, "labels": labels}
+
+    def _tensorize_batch(self, examples: List[torch.Tensor]) -> torch.Tensor:
+        length_of_first = examples[0].size(0)
+        are_tensors_same_length = all(x.size(0) == length_of_first for x in examples)
+        if are_tensors_same_length:
+            return torch.stack(examples, dim=0)
+        else:
+            if self.tokenizer._pad_token is None:
+                raise ValueError(
+                    "You are attempting to pad samples but the tokenizer you are using"
+                    f" ({self.tokenizer.__class__.__name__}) does not have one."
+                )
+            return pad_sequence(examples, batch_first=True, padding_value=self.tokenizer.pad_token_id)
+
+    def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+        """
+        The mask is determined by iteratively performing the following steps until the end of the sequence:
+            1. Sample an n-gram length (n) from [1, ..., max_gram] (L in the paper)
+            2. Reserve a context of length of n-gram length / mlm_probability (In the paper, context size is K*n, where K is close to the inverse of mlm_probability)
+            3. Sample a starting point and mask n tokens continuously from then
+        """
+
+        if self.tokenizer.mask_token is None:
+            raise ValueError(
+                "This tokenizer does not have a mask token which is necessary for masked language modeling. Remove the --mlm flag if you want to use this tokenizer."
+            )
+
+        labels = inputs.clone()
+        # Creating the mask and target mapping tensors
+        masked_indices = torch.full(labels.shape, 0, dtype=torch.bool)
+        target_mapping = torch.zeros((labels.size(0), labels.size(1), labels.size(1)), dtype=torch.float)
+
+        for i in range(labels.size(0)):
+            cur_len = 0
+            max_len = labels.size(1)
+            # Since we're replacing non-masked (non-functional) tokens with -100 in the labels tensor,
+            # the i-th label corresponds to the i-th token.
+            target_mapping[i] = torch.eye(labels.size(1))
+            while cur_len < max_len:
+                n = torch.randint(1, self.max_gram + 1, (1,)).item()
+                ctx_size = int(n / self.mlm_probability)
+                start = cur_len + torch.randint(ctx_size - n + 1, (1,)).item()
+                end = start + 1
+                while end < max_len and end < start + n:
+                    end += 1
+                masked_indices[i, start:end] = 1
+                cur_len += ctx_size
```

Comment:
Is that equivalent to this line in the original code? https://github.com/zihangdai/xlnet/blob/bbaa3a6fa0b3a2ee694e8cf66167434f9eca9660/data_utils.py#L381 ?

Reply:
But it makes sense!

Reply:
Yes, I think so.